src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2019,7 +2028,91 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
                 throw new ArgumentNullException(nameof(nuGetProjectContext));
             }
 
+            token.ThrowIfCancellationRequested();
+
+            var results = new List<NuGetProjectAction>();
+
+            // BuildIntegratedNuGetProject type projects are now supports parallel preview action for faster performance.
+            var buildIntegratedProjectsToUpdate = new List<BuildIntegratedNuGetProject>();
+            var otherTargetProjectsToUpdate = new List<NuGetProject>();
+
+            foreach (var proj in nuGetProjects)
+            {
+                if (proj is BuildIntegratedNuGetProject buildIntegratedNuGetProject)
+                {
+                    buildIntegratedProjectsToUpdate.Add(buildIntegratedNuGetProject);
+                }
+                else
+                {
+                    otherTargetProjectsToUpdate.Add(proj);
+                }
+            }
+
+            if (buildIntegratedProjectsToUpdate.Count != 0)
+            {
+                // Run build integrated project preview for all projects at the same time
+                // There we can do evaluation DependencyGraphRestoreUtility.PreviewRestoreProjectsAsync in batch projects instead of one by one projects.
+                // So for BuildIntegratedNuGetProject type we share same method 'PreviewBuildIntegratedProjectsActionsAsync' for both install and uninstall actions.
+                var uninstallActions = await PreviewBuildIntegratedNuGetProjectsUninstallPackageInternalAsync(
+                buildIntegratedProjectsToUpdate,
+                packageId,
+                nuGetProjectContext,
+                token);
+
+                results.AddRange(uninstallActions);
+            }
+
+            foreach (var target in otherTargetProjectsToUpdate)
```

Comment:
@nkolev92 Is it ok to run this parallel for non-builtIntegrated projects. Actually we don't do it for install action for this types..

Reply:
We don't do it for non-buildintegrated projects, so we shouldn't do it for uninstall either.

Reply:
Ok. Removed.

Reply:
Removed.